client/src/lib.rs

```diff
@@ -114,6 +114,16 @@ mod blocking;
 #[cfg(feature = "async")]
 mod nonblocking;
 
+macro_rules! maybe_async {
+    ($e:expr) => {{
+        #[cfg(feature = "async")]
+        let result = $e.await;
+        #[cfg(not(feature = "async"))]
+        let result = $e;
+        result
+    }};
+}
```

Comment:
The API looks much nicer now, but this macro usage introduces a new problem. Our common internal logic (in `lib.rs`) is async, but this macro uses blocking code based on feature activation. Executing blocking code in async contexts should be avoided, especially if the code is coming from a library. Here is a quote from one of the `tokio` maintainers:

> Async code should never spend a long time without reaching an `.await`.

Full post: https://ryhl.io/blog/async-what-is-blocking/

It would be better to keep the API the same (or similar) and only make internal changes hidden from users to fix this issue.

Reply:
The issue is that in your latest change, you're using blocking code inside async functions. Async functions shouldn't block as I mentioned above.